src/check.rs

```diff
@@ -94,42 +195,183 @@ fn check_http_url(url: &Url, ctx: &CheckContext) -> Result<(), CheckError> {
         }
     }
 
-    let resp = ureq::head(url.as_str()).call();
-    if resp.synthetic() {
-        Err(CheckError::Http(Box::new(HttpError::Fetch(
-            url.clone(),
-            resp.into_synthetic_error().unwrap(),
-        ))))
-    } else if resp.ok() {
-        Ok(())
+    // only if the URL contains a fragment we need to fetch the body via
+    // GET, otherwise we can use HEAD
+    if url.fragment().is_none() {
+        let resp = ureq::head(url.as_str()).call();
+
+        handle_response(url, resp).map(|_: ureq::Response| ())
     } else {
-        Err(CheckError::Http(Box::new(HttpError::UnexpectedStatus(
-            url.clone(),
-            resp,
-        ))))
+        // the URL might contain a fragment, in that case we need to check if
+        // the fragment exists, this issues a GET request
+        check_http_fragment(url, url.fragment().unwrap())
+    }
+}
+cached_key_result! {
+    CHECK_HTTP: SizedCache<(String, String), ()> = SizedCache::with_size(100);
+    Key = { (url.to_string(), fragment.to_string()) };
```

Comment:
Again, let's cache the whole file, not just the specific fragment. That should have substantially fewer HTTP requests.